apps/api/v2/src/ee/overlay-calendars/services/overlay-calendars.service.ts

```diff
@@ -0,0 +1,77 @@
+import { Calendar } from "@/ee/overlay-calendars/inputs/get-overlay-calendars-busy-times.input";
+import { CredentialsRepository } from "@/modules/credentials/credentials.repository";
+import { Injectable, InternalServerErrorException, UnauthorizedException } from "@nestjs/common";
+import { User } from "@prisma/client";
+import { DateTime } from "luxon";
+
+import { getBusyCalendarTimes } from "@calcom/platform-libraries";
+import { ValueType } from "@calcom/platform-types";
+
+@Injectable()
+export class OverlayCalendarsService {
+  constructor(private readonly credentialsRepository: CredentialsRepository) {}
+
+  async getUniqCalendarCredentials(calendarsToLoad: Calendar[], userId: User["id"]) {
+    const uniqueCredentialIds = Array.from(new Set(calendarsToLoad.map((item) => item.credentialId)));
+    const credentials = await this.credentialsRepository.getUserCredentialsByIds(userId, uniqueCredentialIds);
+
+    if (credentials.length !== uniqueCredentialIds.length) {
+      throw new UnauthorizedException("These credentials do not belong to you");
+    }
+
+    return credentials;
+  }
+
+  async getCalendarsWithCredentials(
+    credentials: ValueType<ReturnType<typeof this.credentialsRepository.getUserCredentialsByIds>>,
```

Comment:
Wherever `ValueType<ReturnType...` is used - shall that be created as a custom type in the `credentialsRepository` in this case and then have `credentials: Credentials` ?